[VDO-5804] Skip corruption check for invalid recovery journal blocks

slegendr opened the pull request:
I was not able to reproduce the issue we've been seeing in nightly in the TornWrite tests due to its spurious nature. However, I believe the issue may be due to us checking for invalid blocks after we check for corruption. As such, I have moved around those checks so that we first verify we have a valid recovery journal block before we check if it is corrupted.

These modifications made it possible to simplify the parse_journal_for_recovery() function and eliminate the first loop over the recovery journal blocks that had been added in PR180.

lorelei-sakai reviewed (changes requested):
The code changes looks good. I'll probably squash this commit with the second one from PR 180 for going upstream. It would be nice if we had a testcase that could tell the difference between this code and the prior version, though.

slegendr commented:
> The code changes looks good. I'll probably squash this commit with the second one from PR 180 for going upstream. It would be nice if we had a testcase that could tell the difference between this code and the prior version, though.

I added a commit with tests of basic recovery, as well as of a journal with a block that has a corrupt header and a hole due to a bad sequence number. The head and sequence numbers are not realistic for the non-wrapped journal patterns I added, as there is not sufficient documentation on how to create these patterns.

lorelei-sakai commented on src/c++/vdo/tests/RecoveryJournal_t2.c:
```diff
@@ -112,6 +112,21 @@ const SectorPattern noSectors[VDO_SECTORS_PER_BLOCK] = {
+ * The reap head is 1 and the highest sequence number is 5.
+ **/
+static BlockPattern basicRecoveryPattern[JOURNAL_BLOCKS] = {
+  { 0, 0, GOOD_COUNT, USE_NONCE, FULL_BLOCK, false, normalSectors },
```
I'm probably call this "fullBlocksPattern" or basicFullBlocksPattern" for this. "basic" is a bit generic as a description, and "recovery" is implicit in the test. The pattern itself looks good.

slegendr replied:
Fixed.

lorelei-sakai commented on src/c++/vdo/tests/RecoveryJournal_t2.c:
```diff
@@ -321,6 +336,23 @@ static BlockPattern corruptHeaderPattern[JOURNAL_BLOCKS] = {
+  { 1, 2, GOOD_COUNT, USE_NONCE,    FULL_BLOCK, true,  normalSectors },
+  { 1, 3, GOOD_COUNT, USE_NONCE,    FULL_BLOCK, true,  normalSectors },
+  { 1, 8, GOOD_COUNT, BAD_METADATA, FULL_BLOCK, false, normalSectors },
+  { 1, 5, GOOD_COUNT, USE_NONCE,    FULL_BLOCK, false, normalSectors },
```
Instead of "1, 8", use "9, 12".

slegendr replied:
Fixed.

lorelei-sakai commented on src/c++/vdo/tests/RecoveryJournal_t2.c:
```diff
@@ -453,6 +485,12 @@ static void attemptRebuild(CorruptionType  corruption,
 
+/**********************************************************************/
+static void testBasicRecovery(void)
+{
```
Similar to the above, maybe "testRebuildFullBlocks"?

slegendr replied:
Fixed

lorelei-sakai reviewed (changes requested):
just some very minor things, new testcases look good.

lorelei-sakai commented on src/c++/vdo/tests/RecoveryJournal_t2.c:
```diff
@@ -321,6 +336,23 @@ static BlockPattern corruptHeaderPattern[JOURNAL_BLOCKS] = {
 
+/**
+ * A non-wrapped journal that has been corrupted by the old header
+ * version being written in block 4, along with a hole in block 4 due
```
Test comment is technically wrong (sorry), it's a wrapped journal now  (Could just say "A journal")
I'd also dro reference s to "block 4" specfically,  just say "... corrupted by a hole with the old header version. (Being a hole implies the older sequence number.)

slegendr replied:
Fixed.

lorelei-sakai reviewed (approved):
Great. Looks good.
